File: e2e/fixtures/byte_echo/go/main.go
package main

import (
	"crossfuzz/harness/go"
)

// target echoes the input. The byte-category switch is there only so the
// fuzzer has a few branches to discover — without it, coverage would saturate
// on the first input and the corpus would never grow, defeating the
// "paths discovered" assertion in the e2e harness tests.
func target(data []byte) ([]byte, error) {
	out := make([]byte, len(data))
	for i, b := range data {
		switch {
		case b < 0x20:
			out[i] = b
		case b < 0x40:
			out[i] = b
		case b < 0x60:
			out[i] = b
		case b < 0x80:
			out[i] = b
		default:
			out[i] = b
		}
	}
	return out, nil
}

func main() {
	crossfuzz.Fuzz(target)
}

File: e2e/fixtures/byte_echo/rust/src/main.rs
fn echo(data: &[u8]) -> Vec<u8> {
    let mut out = vec![0u8; data.len()];
    for (i, &b) in data.iter().enumerate() {
        if b < 0x20      { out[i] = b; }
        else if b < 0x40 { out[i] = b; }
        else if b < 0x60 { out[i] = b; }
        else if b < 0x80 { out[i] = b; }
        else             { out[i] = b; }
    }
    out
}

fn main() {
    crossfuzz::fuzz(|input| Ok(echo(input)), Default::default());
}
